src/server/fs/lib/fs-manager.js

```diff
@@ -1249,30 +1229,78 @@ router.post('/webida/api/fs',
+        var fsid;
+
+        // rollback function
+        var rollback = function (err, result) {
+            var msg;
+            switch (state) {
+            case STATE.ASSIGNPOLICY:
+                var policy = result;
+                msg = msg || 'addNewFS assignPolicy fail';
+                authMgr.deletePolicy(policy.pid, user.token, function (err) {
+                    if (err) {
+                        logger.debug('rollback: deletePolicy fail', err);
+                    } else {
+                        logger.debug('rollback: deletePolicy done');
+                    }
+                });
+                // falls through
+            case STATE.CREATEPOLICY:
+                msg = msg || 'addNewFS createDefaultFSPolicy fail';
+                deleteFS(user, fsid, function (err) {
+                    if (err) {
+                        logger.debug('rollback: deletFS fail', err);
+                    } else {
+                        logger.debug('rollback: deletFS done');
+                    }
+                });
+                // falls through
+            case STATE.ADDNEWFS:
+                msg = msg || 'addNewFS fail';
+                break;
+            default:
+                msg = msg || 'addNewFS unknown fail';
+            }
+            logger.info(msg, err, result);
+            return res.sendfail(err, 'Failed to create new filesystem');
+        };
+
+        /* create fs */
+        async.waterfall([
+            function (cb) {
+                state = STATE.ADDNEWFS;
+                addNewFS(user, owner, cb);
+            },
+            function (fsinfo, cb) {
+                fsid = fsinfo.fsid;
+                state = STATE.CREATEPOLICY;
+                policyRule.resource = ['fs:' + fsid + '/*'];
+                authMgr.createPolicy(policyRule, user.token, _.partialRight(cb, fsinfo));
+            },
+            function (policy, fsinfo, cb) {
+                if (!policy) {
+                    return cb(new ServerError('createPolicy failed ' +
+                            JSON.stringify(policyRule)));
                 }
-                logger.debug('assignPolicy', policy);
-                authMgr.assignPolicy(user.uid, policy.pid, user.token, function(err) {
+                state = STATE.ASSIGNPOLICY;
+                authMgr.assignPolicy(owner, policy.pid, user.token, function (err) {
                     if (err) {
-                        logger.info('addNewFS assignPolicy fail', err);
-                        return res.sendfail(err, 'Failed to create new filesystem');
+                        return cb(err, policy);
+                    } else {
+                        return cb(null, fsinfo);
                     }
-
-                    logger.info('addNewFS success', fsinfo);
-                    res.sendok(fsinfo);
                 });
-            });
+            },
+        ], function (err, result) {
+            if (err) {
+                return rollback(err, result);
```

Comment:
I think that it is more safe to pass state value to a parameter of this rollback function.